qiskit/circuit/controlflow/builder.py

```diff
@@ -0,0 +1,378 @@
+    """
+
+    _directive = True
+
+    @abc.abstractmethod
+    def concrete_instruction(
+        self, qubits: FrozenSet[Qubit], clbits: FrozenSet[Clbit]
+    ) -> Tuple[Instruction, Tuple[Qubit, ...], Tuple[Clbit, ...]]:
+        """Get a concrete, complete instruction that is valid to act over all the given resources.
+
+        The returned resources may not be the full width of the given resources, but will certainly
+        be a subset of them; this can occur if (for example) a placeholder ``if`` statement is
+        present, but does not itself contain any placeholder instructions.  For resource efficiency,
+        the returned :obj:`.IfElseOp` will not unnecessarily span all resources, but only the ones
+        that it needs.
+
+        Any condition added in by a call to :obj:`.Instruction.c_if` will be propagated through, but
+        set properties like ``duration`` will not; it doesn't make sense for control-flow operations
+        to have pulse scheduling on them.
+
+        Args:
+            qubits: The qubits the created instruction should be defined across.
+            clbits: The clbits the created instruction should be defined across.
+
+        Returns:
+            Instruction: a full version of the relevant control-flow instruction.  This is a
+            "proper" instruction instance, as if it had been defined with the correct number of
+            qubits and clbits from the beginning.
+        """
+        raise NotImplementedError
+
+    @abc.abstractmethod
+    def placeholder_resources(self) -> Tuple[Tuple[Qubit, ...], Tuple[Clbit, ...]]:
+        """Get the qubit and clbit resources that this placeholder instruction should be considered
+        as using before construction.
+
+        This will likely not include *all* resources after the block has been built, but using the
+        output of this method ensures that all resources will pass through a
+        :meth:`.QuantumCircuit.append` call, even if they come from a placeholder, and consequently
+        will be tracked by the scope managers.
+
+        Returns:
+            A 2-tuple of the quantum and classical resources this placeholder instruction will
+            certainly use.
+        """
+        raise NotImplementedError
+
+    def _copy_mutable_properties(self, instruction: Instruction) -> Instruction:
+        """Copy mutable properties from ourselves onto a non-placeholder instruction.
+
+        The mutable properties are expected to be things like ``condition``, added onto a
+        placeholder by the :meth:`.c_if` method.  This mutates ``instruction``, and returns the same
+        instance that was passed.  This is mostly intended to make writing concrete versions of
+        :meth:`.concrete_instruction` easy.
+
+        Args:
+            instruction: the concrete instruction instance to be mutated.
+
+        Returns:
+            The same instruction instance that was passed, but mutated so its ``condition`` matches
```

Comment:
I was actually going for the opposite approach here - I wanted it to be specific, so you knew exactly what modifications would be performed if you were reading the docs (I have plans to make a new "devdocs" portion of the docs, where I _would_ add these API docs).

I'm not 100% sold one way or the other, though - I totally buy the same arguments you make, because I've seen enough out-of-date documentation while working on libraries.

Reply:
I'm fine with either as well, but it seems odd that the summary text offers `condition` as an example, but the return text makes it explicit. If it's helpful to specify each modification explicitly (for these future devdocs 😄), perhaps an itemized list in the summary would be easier to keep in sync and easier for readers to discover.

Reply:
Yeah, sounds good.  If we get out of sync later, we can always revisit the documentation as soon as we notice!  Done in c0b8cba8e.